packages/pigeon/lib/dart_generator.dart

```diff
@@ -263,14 +264,26 @@ void _writeFlutterApi(
   });
 }
 
-String _addGenericTypes(String dataType, String nullTag) {
-  switch (dataType) {
+String _flattenTypeArguments(List<TypeArgument> args, String nullTag) {
+  return args
+      .map((TypeArgument arg) => arg.typeArguments == null
+          ? '${arg.dataType}$nullTag'
+          : '${arg.dataType}<${_flattenTypeArguments(arg.typeArguments!, nullTag)}>$nullTag')
+      .reduce((String value, String element) => '$value, $element');
+}
+
+String _addGenericTypes(Field field, String nullTag) {
```

Comment:
A declaration comment would be useful here. E.g., it's non-obvious to me what "nullTag" means, especially since it seems to be applied to recursively in a way that seems surprising to me—I'm not sure why a list being nullable and its contents being nullable are conceptually a single argument since there are perfectly valid types that can't express.

Reply:
Added a docstring.  I don't know if this is the best place to document `nullTag` it's used all throughout the Dart generator to essentially represent if you are generating nullsafe code or not.  I imagine for people looking at it they might confuse it with actually having some control over nullsafety.  The use cases that are accepted should be asserted in the front end compiler until we have full nullsafety support.  Clear as mud?